### ISSUE TITLE
Trap within function doesn't catch first signal

### ISSUE THREAD

**Issue by alphapapa** [fish-shell/fish-shell#2036]
I have this shell function:

``` fish
function watchping
    set tempfile (mktemp)
    echo "tempfile: $tempfile"

    trap "rm -f $tempfile; trap SIGINT; echo untrapped" SIGINT
    or begin
        echo "Couldn't trap"
        return 1
    end

    while true
        ping -c1 $argv >$tempfile ^&1

        if test $status -eq 0
            echo -en (date)": "(cat $tempfile | awk 'BEGIN { ORS=" " } NR==2 { print $8 "ms" } NR==5 { print $6 "packet loss" }' | sed 's/time=//')"\n"
        else
            echo (date)": "(cat $tempfile)
        end

        sleep 5
    end
end
```

When I use Ctrl+C, `ping` terminates, and the function ends, and a prompt appears.  But the trap doesn't run (removing the tempfile and echoing "untrapped") until I hit Ctrl+C a second time.

Is this a bug?  It seems like one, because since the function ends, it seems like the trap within the function should activate.

```
$ fish --version
fish, version 2.1.2-1241-g9f8cec7
```

Thanks.

**Comment by ridiculousfish** [comment]
Looks like `trap` is totally busted on ToT. The example from `man trap` is `trap 'status --print-stack-trace' SIGUSR1` and this spews some nonsense.

**Comment by ridiculousfish** [comment]
Equally busted in 2.1.2 on OS X - maybe due to getopt differences.

**Comment by ridiculousfish** [comment]
It looks like our getopt hacks result in double-argument splitting. If the command to execute has no spaces it works OK.

**Comment by ridiculousfish** [comment]
@alphapapa Trapping signals only catches signals delivered to the shell process, not its children. When `ping` is in the foreground, control-C delivers SIGINT to `ping` and not to the shell, so the trap doesn't fire. AFAIK all shells work this way.

If you want to catch when a command ends, you can define a function with `--on-process-exit`. Does that help?

**Comment by alphapapa** [comment]
Thanks for looking into this.

> When ping is in the foreground, control-C delivers SIGINT to ping and not to the shell, so the trap doesn't fire. AFAIK all shells work this way.

I'm far from an expert, but I'm not sure this is universally true.  For example, I have a Bash script that uses `youtube-dl` to download videos and launches VLC to play them.  I trap INT and TERM and EXIT, and when I hit Ctrl+C in the terminal while VLC is running, VLC receives the signal and exits, and the script also receives the signal, so its trap executes.  So it seems like Fish should also receive the signals that its child processes receive.

> If you want to catch when a command ends, you can define a function with --on-process-exit. Does that help?

I'm not sure it does here, because here `ping` is being executed repeatedly; I want to catch when the function itself exits.  I guess I could do this as a script instead of a shell function, but it seems like the trapping issue would remain, because...

In this function, `ping` only runs for a very short time, and then the function sleeps, so the function spends the vast majority of its time sleeping.  In fact, when I hit Ctrl+C in this function, `sleep` is what's running, not `ping`.  I just tested it, making sure that it was sleeping when I interrupted it, and this issue remains, with my having to interrupt it twice before the trap executes.

...I was thinking `sleep` was a builtin, but it's not.  Anyway, it seems like the shell needs to receive the signals its child processes do, because this case of interrupting a script that's currently waiting for `sleep` to return must be a common one, and having to Ctrl+C twice for a trap to execute would seem like a significant bug, leading to cleanup functions never running.

**Comment by ridiculousfish** [comment]
Looks like you're right regarding bash, my mistake.

**Comment by ridiculousfish** [comment]
It seems fish has always worked this way - `--on-signal` handlers only fire for signals delivered to fish, not subprocesses. I don't know how to evaluate which way is better. If anyone feels strongly about this please speak up.

**Comment by alphapapa** [comment]
Respectfully, I strongly think that signals need to pass through to fish.  :)  For scripting, if signals don't pass through to the fish script, then it creates a race condition, trying to get the signal sent to both the current subprocess and the fish process before fish executes the next line in the script.  This could continue for several lines of the script before the script is finally interrupted.  If fish doesn't pass the signal through, it would seem to make trapping far less useful, and make it difficult to use fish for scripting, period.  Also, for consistency with other shells, it should behave the same way.

At the moment I can't think of any situations in which one wouldn't want the signals to pass through.  Even if one wanted to only terminate subprocesses, that could be handled by using variables to maintain state and a function to either exit the script or allow execution to continue.  Or an `--ignore-subprocess` flag could be added to `trap` to ignore signals passed to subprocesses, but that shouldn't be the default behavior, because it makes trap not very useful.

Thanks.

**Comment by quartsize** [comment]
In Bash, I'd gotten used to writing

    trap 'exit' SIGINT
    for host in ...; do something-to $host; done

so that one Ctrl-C will stop the whole loop, even if it's in the middle of something-to. I'd like to be able to do something similar with fish, which letting `--on-signal` fire for signals sent to subprocesses would accomplish.

**Comment by alphapapa** [comment]
I've just become further confused about this issue.  For example:
```fish
#!/usr/bin/env fish

function cleanup --on-signal INT --on-process-exit %self
    echo "Cleaning up..."
end

trap cleanup INT TERM

sleep 5 &
cat

echo "Exiting..."
```
I have tried this script with and without `cat`, and with and without putting `sleep` into the background.  No matter what I do, I can't get the `cleanup` function to run when I press Ctrl+C.  But this Bash script works fine:
```bash
#!/bin/bash

function cleanup {
    echo "Cleaning up..."
}

trap cleanup INT TERM

sleep 5 &
cat

echo "Exiting..."
```
What am I doing wrong?

**Comment by michaelperrygoodman** [comment]
I would also like to know why this isn't working:

`function _exit
  echo '[ exiting shell ]'
end
trap _exit EXIT
`
which I have also tried to write as

`
function --on-signal EXIT _exit
  echo '[ exiting shell ]'
end
`

The former worked in BASH land but seems to behave differently here.

**Comment by ridiculousfish** [comment]
EXIT is not a signal anywhere :)

Try:

    function _exit  --on-process-exit %self
        # stuff
    end

note that `echo` may not work in stuff since stdout may already be closed.

**Comment by krader1961** [comment]
@ridiculousfish: executing  `trap 'echo wtf' EXIT` works for me using a fish built from github head. Fish is automatically converting it into something like this:

```fish
function __trap_handler_EXIT --on-process-exit 1008
	echo wtf;
end
```

However, your point about stdout possibly already being closed is worth keeping in mind. 

@michaelperrygoodman: You didn't really type that `function _exit...` all on one line without semicolons did you? Because that definitely will not work. Also, after defining that trap what do you see when you type just `trap` to list the currently defined traps? Lastly, try replacing that echo with something like `touch /tmp/pid_(echo %self)_exited` then check if that file was created.

**Comment by zanchey** [comment]
> You didn't really type that function _exit... all on one line without semicolons

No - just a problem with the formatting which I've fixed (it was clear in the plaintext version - who gets HTML mail anyway?)

**Comment by krader1961** [comment]
> who gets HTML mail anyway?

Something like 98% of Internet users I would guess. Which makes me sad, especially when combined with top posting. So, did you just decide to fix the question or are you also posting as @michaelperrygoodman? Regardless, does the `trap _some_function_or_command EXIT` work for you?

One of these days I should setup a virtual machine with the public 2.2.0 release so I can verify if these types of issues have different behavior with that version.

**Comment by michaelperrygoodman** [comment]
Yes that trap _exit EXIT has been working for me and I assume its a bash thing and not a signal thing

**Comment by maxnordlund** [comment]
I not sure if this belongs here, but I can't seem to find a way to trap function return. That is, do some cleanup when a function returns, regardless if it was interrupted somewhere in the middle.

```fish
function foo
  set -l dir (mktemp -d temporary_stuff)
  function cleanup --on-function-return --inherit-variable dir
    rm -rf $dir
    function --erase cleanup # Should be done automatically?
  end

  # stuff with foo
end
```

**Comment by krader1961** [comment]
@maxnordlund: There is no --on-function-return option for the function command. I'd recommend opening an enhancement issue as I couldn't find an open issue asking for such a feature and it's orthogonal to this issue. It's also unclear whether your proposal is the best way to solve this issue. For example would a "try; ...; except" style block be better and solve other problems? Or would it be better to add define foo like `function foo --on-return=func_name` and have the named function automatically inherit the scope of "foo"?

@ridiculousfish: I agree with @alphapapa that fish should behave like bash and zsh in this regard. How difficult that will be to achieve is TBD as it is likely to affect tty management (something that needs reworking for other reasons) as well as fish's job control. Also, `trap` is currently not scoped to the current block which seems like something that should also be changed (obviously file scope == global if the file is sourced outside of a function).

**Comment by maxnordlund** [comment]
Yes I know it doesn't exist, but wasn't sure if it was different enough to warrant it's own issue, as I belive it is not orthogonal, rather slightly angled to this one.

Using `try`/`catch`/`finally` would not be the quite same, and would mean that every function that needs some cleanup needs to wrap its entire body in an `try` statement. It also confuses events/signals with exceptions, which may or may not be such a good idea.

The extra flag to `function` does sound like a nicer alternative.

**Comment by pirate** [comment]
For anyone stumbling across this via google, you can accomplish something similar to `trap 'cmd' SIGINT` using:
```bash
function long_running_thing --description 'sleep for $argv seconds, supports ctrl-c mid-sleep'
    set time_to_sleep $argv
    
    # set up a handler to be triggered if long_running_thing is cancelled mid-run
    function on_exit --on-job-exit %self --inherit-variable time_to_sleep
        functions -e on_exit  # erase to prevent recursively calling on_exit when this function exits
        echo "Cancelled waiting for $time_to_sleep seconds"
    end

    # run the task
    sleep $time_to_sleep
    echo "Finished waiting for $time_to_sleep seconds"
    
    # erase the on_exit handler before exiting successfully
    functions -e on_exit
end
```

**Comment by alphapapa** [comment]
@pirate That's a very clever hack!  Thanks for sharing.  Unfortunately it doesn't seem to work for me:

```fish
#!/usr/bin/env fish

function dosleep

    function interrupted --on-job-exit %self
        functions -e interrupted
        echo INTERRUPTED
    end

    sleep 50
    echo "Sleep over"

    functions -e interrupted
end

echo one
dosleep
echo two
```
Running that and pressing ^C gives this output:
```
./testtrap.fish 
one
^C
```
Using fish, version 2.5.0-305-ge7d6864.

Tried several combinations of `--on-process-exit`, `--on-job-exit`, `%self`, `%last`, `%sleep`, `(jobs -glp)`, but nothing works.  The `sleep` command gets interrupted and the script exits immediately without running the function.

I hope this can be fixed.  It's a major hurdle to using Fish to write interactive scripts.  I need a script to run a background process and run a command when the background process is interrupted, but this seems impossible with Fish, so I have to use Bash.

**Comment by pirate** [comment]
Yeah @alphapapa it stopped working for me too, I ended up implementing my own backgrounding job control for swift using these functions:
```fish
function quiet --description 'Run a command with all stdout and stderr silenced'
    if not count $argv > /dev/null
        return 0
    end

    # quote every argument so spacing is preserved
    set evalstr (string escape -- $argv)
    # eval the command with output silenced
    eval "$evalstr 2>&1 > /dev/null" 2>&1 > /dev/null; and return 0; or return 1
end

function background --description "Display a process's stdout while it's running in the background"
	bash -c "fish -c '""$argv""' &"
end

function fork --description "Run a process in the background without displaying its stdout or stderr"
    # calls /usr/bin/fork script which contains the following:
    # !/bin/bash
    # /usr/local/bin/fish -c "$*" > ~/tmp/fish_forks/last.log 2>&1 &
    # pid=$!
    # mv ~/tmp/fish_forks/last.log ~/tmp/fish_forks/$pid.log
    # disown $pid
end

function signal --description 'Signal a process thats waiting on a semaphore'
    quiet rm ~/tmp/fish_signals/"$argv"
    touch ~/tmp/fish_signals/"$argv"
end

function await --description 'Wait for a specified signal from another process'
    set -l signal_name "$argv"
    quiet mkdir -p ~/tmp/fish_signals
    quiet rm ~/tmp/fish_signals/"$signal_name" # clear out old semaphore
    echo %self > ~/tmp/fish_signals/"$signal_name".waiter

    # wait for the semaphore to be set
    while not test -e ~/tmp/fish_signals/"$signal_name"
        sleep 0.05
    end
    rm ~/tmp/fish_signals/"$signal_name".waiter
    sleep 0.05

    if test -s ~/tmp/fish_signals/"$signal_name"
        cat ~/tmp/fish_signals/"$signal_name"
    else
        false
    end
end
```

Example usage:
```fish
function do_task
    background "await sleep_done"
    sleep 10
    signal sleep_done
end

function cancellable_task
    function on_ctrl_c --on-job-exit %self
        functions -e on_ctrl_c
        kill (ps ax | grep -v grep | grep "await sleep_done" | awk '{print $1}')
        echo "[X] Cancelled"
    end

    echo "Starting task"
    do_task
    echo "Finished task"  # never reached if Ctrl+C is pressed
    functions -e on_ctrl_c
end
```

Notice how you have to define the cancel handler one level up from the actual task function, it's a bit of a pain but it's the only way I could get proper background task management working.  I have lots of functions built around the `background`, `await`, `signal` trio, including much more complex use cases.  Also check out [progress_await.fish](https://github.com/pirate/fish-functions/blob/master/progress_await.fish) for a progress bar while waiting.

**Comment by alphapapa** [comment]
Thanks, @pirate, I will have to study this carefully.  :)

**Comment by Smarre** [comment]
This still happens with 3.0.0. It also seems to affect --on-event fish_exit handler (if it’s supposed to be executed when script is terminated with ctrl+c while a process is running in it).

**Comment by GeneralDenmark** [comment]
> For anyone stumbling across this thread, I managed to accomplish trapping Ctrl-C's mid-function using:
> 
> ```shell
> function long_running_thing --description 'sleep for $argv seconds, supports ctrl-c mid-sleep'
>     set time_to_sleep $argv
> 
>     # set up a handler to be triggered if long_running_thing is cancelled mid-run
>     function on_premature_exit --on-job-exit %self --inherit-variable time_to_sleep
>         functions -e on_premature_exit  # erase to prevent recursively calling itself when it exits
>         echo "Cancelled waiting for $time_to_sleep seconds"
>     end
> 
>     # run the task
>     sleep $time_to_sleep
>     echo "Finished waiting for $time_to_sleep seconds"
> 
>     # erase the on_exit handler before exiting successfully
>     functions -e on_premature_exit
> end
> ```

This works again in Fish v.3.1.2, or well "_it works on my machine_" :tm:

**Comment by faho** [comment]
I believe this was solved in #6649. At least I can't reproduce it anymore.
